src/miner.cpp

```diff
@@ -286,7 +290,23 @@ class AddOutputsToCoinbaseTxAndSign : public boost::static_visitor<>
     }
 };
 
-CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddress& minerAddress)
+void CoinbaseTransaction(const CChainParams& chainparams, CMutableTransaction& tx, CAmount nFees, const MinerAddress& minerAddress, int nHeight)
+{
+        tx = CreateNewContextualCMutableTransaction(chainparams.GetConsensus(), nHeight);
+        tx.vin.resize(1);
+        tx.vin[0].prevout.SetNull();
+        // Set to 0 so expiry height does not apply to coinbase txs
+        tx.nExpiryHeight = 0;
+
+        // Add outputs and sign
+        boost::apply_visitor(
+            AddOutputsToCoinbaseTxAndSign(tx, chainparams, nHeight, nFees),
+            minerAddress);
+
+        tx.vin[0].scriptSig = CScript() << nHeight << OP_0;
+}
```

Comment:
This is just a refactor, pulling this code out of `CreateNewBlock()` below. It's called from two places, `CreateNewBlock()` and by `getblocktemplate` RPC handler to pre-compute the N+2 block. This is the code that can take 400 ms (if the output is shielded).